pkg/cmd/roachtest/tests/tpcc.go

```diff
@@ -1906,3 +2150,338 @@ func exportOpenMetrics(
+			c.Run(ctx, option.WithNodes(c.Node(1)), cmd)
+		}
+		// Partitioning is done by the init step, run a short workload to make sure
+		// all ranges are set up correctly. This may not be strictly necessary
+		// (consider removing).
+		t.L().Printf("Step 5 - Partition the database (Done automatically by import)")
+		t.L().Printf("Step 6 - This step intentionally left blank to match public docs")
+		{
+			var cmd string
+			extraArgs := ""
+			if opts.optimized {
+				extraArgs = fmt.Sprintf(" --partitions=%d", crdbNodeCount)
+			}
+			cmd = fmt.Sprintf(
+				"./cockroach1 workload run tpcc %s --warehouses=%d  --duration=1m --workers=1000 --wait=0.0 {pgurl%s}",
+				extraArgs,
+				opts.warehouses,
+				crdbNodes,
+			)
+			t.L().Printf("Step 7 - Allocate partitions\n (%s)", cmd)
+			c.Run(ctx, option.WithNodes(c.Node(1)), cmd)
+			t.L().Printf("Waiting for stable ranges")
+
+			// TODO(baptist): Replace the next 2 steps with replication reports once
+			// they are available.
+			func() {
+				db := c.Conn(ctx, t.L(), 1)
+				defer db.Close()
+				for {
+					var n int
+					require.NoError(t,
+						db.QueryRowContext(
+							ctx,
+							"SELECT count(1) FROM crdb_internal.ranges WHERE array_length(replicas, 1) < 3",
+						).Scan(&n))
+					if n == 0 {
+						t.L().Printf("All ranges fully upreplicated")
+						break
+					}
+					time.Sleep(10 * time.Second)
+				}
+			}()
+			for {
+				// Run all the queries in parallel to find the total pending count.
+				found := make(chan int)
+				for _, nodeID := range crdbNodes {
+					nodeID := nodeID
+					go func() {
+						db := c.Conn(ctx, t.L(), nodeID)
+						defer db.Close()
+						var n int
+						require.NoError(t,
+							db.QueryRowContext(
+								ctx,
+								"SELECT value FROM crdb_internal.node_metrics WHERE name = 'queue.replicate.pending'",
+							).Scan(&n))
+						found <- n
+					}()
+				}
+				var total int
+				// Wait until they have all completed.
+				for range crdbNodes {
+					total += <-found
+				}
+				if total == 0 {
+					t.L().Printf("All ranges in final locations")
+					break
+				}
+				t.L().Printf("Waiting for ranges to move: %d", total)
+				time.Sleep(60 * time.Second)
+			}
+		}
+		{
+			// Reset settings that aren't needed anymore and could impact performance.
+			db := c.Conn(ctx, t.L(), 1)
+			_, _ = db.ExecContext(ctx, `RESET CLUSTER SETTING kv.snapshot_rebalance.max_rate`)
+			require.NoError(t, db.Close())
+		}
+	}
+
+	t.L().Printf("Step 8 - Run the benchmark")
+	// Create all the workers that are used below.
+	var workers []worker
+	for i := 0; i < workloadCount; i++ {
+		start := crdbNodeCount * i / workloadCount
+		end := crdbNodeCount * (i + 1) / workloadCount
+		// Note that this is not always the same for each iteration due to rounding.
+		lists := make([]string, end-start)
+		// Use this to handle rounding well.
+		for j := start; j < end; j++ {
+			lists[j-start] = fmt.Sprint(j)
```

Comment:
Nit: use `strconv.Itoa`; `fmt.Sprint` uses reflection and thus less preferable due to overhead.